examples/ecs/indexing.rs

```diff
@@ -0,0 +1,205 @@
+//! This example demonstrates how to access `Component` data through an `Index`.
+#![allow(clippy::type_complexity)]
+
+use bevy::{
+    ecs::{
+        component::Tick,
+        system::{SystemChangeTick, SystemParam},
+    },
+    prelude::*,
+    utils::{HashMap, HashSet},
+};
+use bevy_internal::ecs::query::ReadOnlyWorldQuery;
+use std::{hash::Hash, marker::PhantomData};
+
+pub trait Indexer {
+    type Input: Component;
+    type Index: Hash + Eq + Clone + 'static;
+
+    fn index(input: &Self::Input) -> Self::Index;
+}
+
+pub struct SimpleIndexer<T>(PhantomData<T>);
+
+impl<T> Indexer for SimpleIndexer<T> where T: Component + Hash + Eq + Clone {
+    type Input = T;
+
+    type Index = T;
+
+    fn index(input: &Self::Input) -> Self::Index {
+        input.clone()
+    }
+}
+
+#[derive(Resource)]
+struct IndexBacking<T, F = (), I = SimpleIndexer<T>> {
+    forward: HashMap<T, HashSet<Entity>>,
+    reverse: HashMap<Entity, T>,
+    last_this_run: Option<Tick>,
+    _phantom: PhantomData<fn(F, I)>,
+}
+
+impl<T, F, I> Default for IndexBacking<T, F, I> {
+    fn default() -> Self {
+        Self {
+            forward: default(),
+            reverse: default(),
+            last_this_run: default(),
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<T, F> IndexBacking<T, F> {
+    fn update(&mut self, entity: Entity, value: Option<T>) -> Option<T>
```

Comment:
i feel this function should not also handle deletes - update(entity, None) = delete is non-obvious

Reply:
For now I had it as a single method because they had so much overlapping behavior. I do agree that this should be a more standard container type, with all the standard methods that would entail.